drivers/tee/optee/optee.c

```diff
@@ -874,15 +1007,93 @@ static int optee_shm_unregister(const struct device *dev, struct tee_shm *shm)
 	return rc;
 }
 
-static int optee_suppl_recv(const struct device *dev, uint32_t func, unsigned int num_params,
+static int optee_suppl_recv(const struct device *dev, uint32_t *func, unsigned int *num_params,
 			    struct tee_param *param)
 {
+	struct optee_driver_data *data = (struct optee_driver_data *)dev->data;
+	struct optee_supp *supp = &data->supp;
+	struct optee_supp_req *req = NULL;
+
+	if (supp->current) {
+		return -EBUSY;
+	}
+
+	while (true) {
+		k_mutex_lock(&supp->mutex, K_FOREVER);
+		if (!sys_dlist_is_empty(&supp->reqs)) {
+			req = (struct optee_supp_req *)sys_dlist_get(&supp->reqs);
+		}
+		k_mutex_unlock(&supp->mutex);
+
+		if (req) {
+			break;
+		}
+
+		k_sem_take(&supp->reqs_c, K_FOREVER);
+	}
+
+	if (*num_params < req->num_params) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&supp->mutex, K_FOREVER);
+	supp->current = req;
+	k_mutex_unlock(&supp->mutex);
+
+	*func = req->func;
+	*num_params = req->num_params;
+	memcpy(param, req->param, sizeof(struct tee_param) * req->num_params);
+
 	return 0;
 }
 
-static int optee_suppl_send(const struct device *dev, unsigned int num_params,
+static int optee_suppl_send(const struct device *dev, unsigned int ret, unsigned int num_params,
```

Comment:
do we really need `num_params`? should you check that we have enough number of items to hold `req->num_params`
as far as I see, it is unused

Reply:
Added check